README


This archive contains the distribution Audio-MPD,
version 1.120610:

  class to talk to MPD (Music Player Daemon) servers

This software is copyright (c) 2007 by Jerome Quelin.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Changelog for Audio-MPD

{{$NEXT}}

2.004     2016-05-20 14:05:55+02:00 Europe/Paris
 - fix rt#99762 - support for mpd on a Unix socket (smcv)

2.003     2016-05-20 14:01:00+02:00 Europe/Paris
 - fix rt#114136 - allow mpd-dynamic to run without a ratings.db

2.002     2016-04-28 20:35:51+02:00 Europe/Paris
 - fix rt#100300 - test support for mpd 0.19 (Debian)

2.001     2016-04-28 20:27:01+02:00 Europe/Paris
 - mpd-dynamic: simplify code & logic
 - new script mpd-dump-ratings

2.000     2014-06-11 14:31:10+02:00 Europe/Paris
 - adapt to new Moose (rt#90682)
 - fix test breakages due to mpd output changes (rt#91082)
 - migrated to @AUthor::JQUELIN

1.120610  2012-03-01 09:31:01 Europe/Paris
 - fix rt#75436 - mpd-rate and mpd-dynamic should fail gracefully when
   no song is playing (florian)
 - fix rt#75437 - fix mpd-rate and mpd-dynamic pod names

1.113280  2011-11-24 15:07:47 Europe/Paris
 - fix rt#68357 - ipv6 support
 - fix rt#71201 - url handlers can vary with mpd installation
 - fix rt#71202 - requiring software mixer for tests
 - tests for all_genres, artists_by_genre (sebastien nobili)
 - better namespace for scripts

1.112670  2011-09-24 11:40:44 Europe/Paris
 - collection: new method all_genres (sebastien nobili)
 - collection: new method artists_by_genre (sebastien nobili)

1.111200  2011-04-30 10:21:31 Europe/Paris
 - removing misleading example in pod

1.110560  2011-02-25 15:28:59 Europe/Paris
 - new command available: outputs

1.102260  2010-08-14 19:21:51 Europe/Paris
 - new script mpd-rate to rate a song / display current rating

1.100430  2010-02-12 12:51:18 Europe/Paris
 - WARNING: no more makefile.pl supplied, only a build.pl
 - sub::exporter no more used internally

1.093170 Fri Nov 13 19:06:54 2009
 - removed no more used prereq class::accessor
 - revamped pod with podweaver
 - introducing audio::mpd::types
 - using moosex::has::sugar

1.092950 Thu Oct 22 19:34:35 2009
 - module is quite stable now, bumping to 1.yydddn version
 - migrated to moose
 - password() method renamed to set_password()
 - no more vars exported, use 'reuse' and 'once' strings for conntype

**

0.19.10 Wed Oct 21 14:20:09 2009
 - update list of static prereqs, not caught by autoprereq
 - bump min version for test-corpus-audio-mpd

0.19.9 Sun Oct 18 11:49:57 2009
 - using test::corpus::audio::mpd for tests (regains mpd 0.15.x compatibility)

0.19.8 Wed Oct 14 18:05:29 2009
 - migrated to dist-zilla (transparent for end-user)

0.19.7 Wed Sep 30 16:56:37 2009
 - fix #50109 problem with newer class::accessor

0.19.6 Sat Feb 14 13:03:22 CET 2009
 - doc update (catched by streamingmeemee)

0.19.5 Mon Jan  5 15:31:29 CET 2009
 - updated to match mpd 0.14 behaviour
 - doc update

0.19.4 Sat Aug 16 10:38:12 CEST 2008
 - doc update
 - listing mpd-dynamic script prereqs as module prereqs

0.19.3 Tue Aug 12 19:04:23 CEST 2008
 - version() reports *protocol* version, not real mpd version
 - fix some tests that failed because of detection method not precise
   enough (timing issues)

0.19.2 Tue Jun 17 19:24:16 CEST 2008
 - mpd-dynamic should lookup ratings with utf8 encoded pathes

0.19.1 Mon Dec  3 18:23:58 CET 2007
 - new method all_playlists() (courtesy of Joey Hess)
 - fix for filenames / playlists beginning with a space (courtesy of
   Joey Hess)

0.19.0 Thu Nov 29 20:10:18 CET 2007
 - changed constructor api: using a hash for options instead of
   positional paraemters
 - new conntype param for constructor to change the way connection is
   handled
 - MPD_HOST env var supports password@host
 - mpd-dynamic daemonizes after mpd connection to trap problems

**

0.18.3 Mon Nov 26 12:51:40 CET 2007
 - kwalitee/cpants release

0.18.2 Tue Nov 13 20:20:59 CET 2007
 - fixing bug 29232 (courtesy of David Bitseff)

0.18.1 Thu May 31 20:57:47 CEST 2007
 - fixing mantis bug #556: support for multiple $path in playlist->add

0.18.0 Wed May 30 12:29:11 CEST 2007
 - moved to perl license
 - using common classes outsourced in audio::mpd::common, namely: item*,
   stats, status and time.
 - new methods AM::Collection::all_songs()

**

0.17.2 Sun Apr  8 15:35:43 CEST 2007
 - introducing AM::Stats for better encapsulation

0.17.1 Sun Apr  8 12:23:53 CEST 2007
 - introducing AM::Time for better encapsulation
 - thus, removed get_time_info() and get_time_format()
 - test coverage up to 100% \o/

0.17.0 Fri Apr  6 16:54:07 CEST 2007
 - introducing AM::Playlist for better segregation of playlist handling
 - completed tests to cover all playlist operations, AM is now 95.1% covered

**

0.16.2 Tue Apr  3 19:02:29 CEST 2007
 - utf-8 support generalized

0.16.1 Mon Apr  2 18:26:17 CEST 2007
 - new internal _cooked_command_*() methods, allowing for better code
   factorization - and thus, better code coverage (100% achieved for
   am::collection, 89.3% overall)

0.16.0 Thu Mar 29 19:04:34 CEST 2007
 - script to listen your music endlessly: mpd-dynamic

 **

0.15.4 Tue Mar 27 18:42:58 CEST 2007
 - AM::Test more robust regarding fake mpd startup
 - AM::Test quicker if no tests performed
 - new methods in AM::Collection:
    . songs_with_filename_partial()
    . songs_by_artist_partial()
    . songs_from_album_partial()
    . songs_with_title_partial()

0.15.3 Wed Mar 21 18:08:06 CET 2007
 - requiring perl 5.8.x minimum

0.15.2 Tue Mar 20 17:24:59 CET 2007
 - fixed bug #25553 (on rt.cpan.org)

0.15.1 Sun Mar 18 19:06:57 CET 2007
 - added Readonly as prereq

0.15.0 Sat Mar 17 17:49:50 CET 2007
 - implementation of Audio::MPD::Collection: this class will act as the entry
   point to look up in mpd's collection
 - thus, following AM methods removed (see AM::Collection for the replacement):

lib/Audio/MPD.pm
use 5.008;
use warnings;
use strict;

package Audio::MPD;
# ABSTRACT: class to talk to MPD (Music Player Daemon) servers

use Audio::MPD::Common::Item;
use Audio::MPD::Common::Stats;
use Audio::MPD::Common::Status;
use Audio::MPD::Common::Output;
use Encode;
use IO::Socket::IP;
use Moose;
use MooseX::Has::Sugar;
use MooseX::SemiAffordanceAccessor;

use Audio::MPD::Collection;
use Audio::MPD::Playlist;
use Audio::MPD::Types;


=attr host

The hostname where MPD is running. Defaults to environment var
C<MPD_HOST>, then to 'localhost'. Note that C<MPD_HOST> can be of
the form C<password@host:port> (each of C<password@> or C<:port> can
be omitted).

=attr port

The port that MPD server listens to. Defaults to environment var
C<MPD_PORT>, then to parsed C<MPD_HOST> (cf above), then to 6600.

=attr password

The password to access special MPD functions. Defaults to environment
var C<MPD_PASSWORD>, then to parsed C<MPD_HOST> (cf above), then to
empty string.

=attr conntype

Change how the connection to mpd server is handled. It should be of a
C<CONNTYPE> type (cf L<Audio::MPD::Types>). Use either the C<reuse>
string to reuse the same connection or C<once> to open a new connection
per command (default).

=cut

has conntype   => ( ro, isa=>'CONNTYPE', default=>'once'  );
has host       => ( ro, lazy_build );
has password   => ( rw, lazy_build, trigger=>sub { $_[0]->ping } );
has port       => ( ro, lazy_build );

has collection => ( ro, lazy_build, isa=>'Audio::MPD::Collection' );
has playlist   => ( ro, lazy_build, isa=>'Audio::MPD::Playlist'   );
has version    => ( rw );

has _socket    => ( rw, isa=>'IO::Socket' );


#--
# initializer & lazy builders

=method new

    my $mpd = Audio::MPD->new( \%opts );

This is the constructor for L<Audio::MPD>. One can specify any of the
attributes (cf above).

=cut

sub BUILD {
    my $self = shift;

    # create the connection if conntype is set to $REUSE
    $self->_connect_to_mpd_server if $self->conntype eq 'reuse';

    # try to issue a ping to test connection - this can die.
    $self->ping;
}

#
# my ($passwd, $host, $port) = _parse_env_var();
#
# parse MPD_HOST environment variable, and extract its components. the
# canonical format of MPD_HOST is passwd@host:port.
#
sub _parse_env_var {
    return (undef, undef, undef) unless defined $ENV{MPD_HOST};
    return ($1, $2, $3)    if $ENV{MPD_HOST} =~ /^([^@]+)\@([^:@]+):(\d+)$/; # passwd@host:port
    return ($1, $2, undef) if $ENV{MPD_HOST} =~ /^([^@]+)\@([^:@]+)$/;       # passwd@host
    return (undef, $1, $2) if $ENV{MPD_HOST} =~ /^([^:@]+):(\d+)$/;          # host:port
    return (undef, $ENV{MPD_HOST}, undef);
}

sub _build_host     { return ( _parse_env_var() )[1] || 'localhost'; }
sub _build_port     { return $ENV{MPD_PORT}     || ( _parse_env_var() )[2] || 6600; }
sub _build_password { return $ENV{MPD_PASSWORD} || ( _parse_env_var() )[0] || '';   }

sub _build_collection { Audio::MPD::Collection->new( _mpd => $_[0] ); }
sub _build_playlist   { Audio::MPD::Playlist  ->new( _mpd => $_[0] ); }


#--
# Private methods


#
# $mpd->_connect_to_mpd_server;
#
# This method connects to the mpd server. It can die on several conditions:
#  - if the server cannot be reached,
#  - if it's not an mpd server,
#  - or if the password is incorrect,
#
sub _connect_to_mpd_server {
    my ($self) = @_;

    # try to connect to mpd.
    my $socket;

    if ($self->host =~ m{^/}) {
        eval q{use IO::Socket::UNIX qw(); 1}
            or die "Could not load IO::Socket::UNIX: $@\n";
        $socket = IO::Socket::UNIX->new($self->host)
            or die "Could not create socket: $!\n";
    }
    else {
        $socket = IO::Socket::IP->new(
            PeerAddr => $self->host,
            PeerPort => $self->port,
        )
        or die "Could not create socket: $!\n";
    }

    # parse version information.
    my $line = $socket->getline;
    chomp $line;
    die "Not a mpd server - welcome string was: [$line]\n"
        if $line !~ /^OK MPD (.+)$/;
    $self->set_version($1);

    # send password.
    if ( $self->password ) {
        $socket->print( 'password ' . encode('utf-8', $self->password) . "\n" );
        $line = $socket->getline;
        die $line if $line =~ s/^ACK //;
    }

    # save socket
    $self->_set_socket($socket);
}


#
# my @result = $mpd->_send_command( $command );
#
# This method is central to the module. It is responsible for interacting with
# mpd by sending the $command and reading output - that will be returned as an
# array of chomped lines (status line will not be returned).
#
# This method can die on several conditions:
#  - if the server cannot be reached,
#  - if it's not an mpd server,
#  - if the password is incorrect,
#  - or if the command is an invalid mpd command.
# In the latter case, the mpd error message will be returned.
#
sub _send_command {
    my ($self, $command) = @_;

    $self->_connect_to_mpd_server if $self->conntype eq 'once';
    my $socket = $self->_socket;

    # ok, now we're connected - let's issue the command.
    $socket->print( encode('utf-8', $command) );
    my @output;
    while (defined ( my $line = $socket->getline ) ) {
        chomp $line;
        die $line if $line =~ s/^ACK //; # oops - error.
        last if $line =~ /^OK/;          # end of output.
        push @output, decode('utf-8', $line);
    }

    # close the socket.
    $socket->close if $self->conntype eq 'once';

    return @output;
}


#
# my @items = $mpd->_cooked_command_as_items( $command );
#
# Lots of Audio::MPD methods are using _send_command() and then parse the
# output as a collection of AMC::Item. This method is meant to factorize
# this code, and will parse the raw output of _send_command() in a cooked
# list of items.
#
sub _cooked_command_as_items {
    my ($self, $command) = @_;

    my @lines = $self->_send_command($command);
    my (@items, %param);

    # parse lines in reverse order since "file:" or "directory:" lines
    # come first. therefore, let's first store every other parameter,
    # and the last line will trigger the object creation.
    # of course, since we want to preserve the playlist order, this means
    # that we're going to unshift the objects instead of push.
    foreach my $line (reverse @lines) {
        my ($k,$v) = split /:\s/, $line, 2;
        $param{$k} = $v;
        next unless $k eq 'file' || $k eq 'directory' || $k eq 'playlist'; # last param of item
        unshift @items, Audio::MPD::Common::Item->new(%param);
        %param = ();
    }

    return @items;
}


#
# my %hash = $mpd->_cooked_command_as_kv( $command );
#
# Lots of Audio::MPD methods are using _send_command() and then parse the
# output to get a list of key / value (with the colon ":" acting as separator).
# This method is meant to factorize this code, and will parse the raw output
# of _send_command() in a cooked hash.
#
sub _cooked_command_as_kv {
    my ($self, $command) = @_;
    my %hash =
        map { split(/:\s/, $_, 2) }
        $self->_send_command($command);
    return %hash;
}

#
# my @list = $mpd->_cooked_command_strip_first_field( $command );
#
# Lots of Audio::MPD methods are using _send_command() and then parse the
# output to remove the first field (with the colon ":" acting as separator).
# This method is meant to factorize this code, and will parse the raw output
# of _send_command() in a cooked list of strings.
#
sub _cooked_command_strip_first_field {
    my ($self, $command) = @_;

    my @list =
        map { ( split(/:\s+/, $_, 2) )[1] }
        $self->_send_command($command);
    return @list;
}


#--
# Public methods

# -- MPD interaction: general commands

=meth_mpd_ctrl ping

    $mpd->ping;

Sends a ping command to the mpd server.

=cut

sub ping {
    my ($self) = @_;
    $self->_send_command( "ping\n" );
}



=meth_mpd_ctrl version

    my $version = $mpd->version;

Return mpd's version number as advertised during connection. Note that
mpd returns B<protocol> version when connected. This protocol version
can differ from the real mpd version. eg, mpd version 0.13.2 is
"speaking" and thus advertising version 0.13.0.

=cut

# sub version {} # implemented as an accessor.



=meth_mpd_ctrl kill

    $mpd->kill;

Send a message to the MPD server telling it to shut down.

=cut

sub kill {
    my ($self) = @_;
    $self->_send_command("kill\n");
}



=meth_mpd_ctrl set_password

    $mpd->set_password( [$password] );

Change password used to communicate with MPD server to C<$password>.
Empty string is assumed if C<$password> is not supplied.

=cut

# implemented by password trigger (from moose)


=meth_mpd_ctrl updatedb

    $mpd->updatedb( [$path] );

Force mpd to recan its collection. If C<$path> (relative to MPD's music
directory) is supplied, MPD will only scan it - otherwise, MPD will
rescan its whole collection.

=cut

sub updatedb {
    my ($self, $path) = @_;
    $path ||= '';
    $self->_send_command("update $path\n");
}


=meth_mpd_ctrl urlhandlers

    my @handlers = $mpd->urlhandlers;

Return an array of supported URL schemes.

=cut

sub urlhandlers {
    my ($self) = @_;
    return $self->_cooked_command_strip_first_field("urlhandlers\n");
}


# -- MPD interaction: handling volume & output

=meth_mpd_output volume

    $mpd->volume( [+][-]$volume );

Sets the audio output volume percentage to absolute C<$volume>.  If
C<$volume> is prefixed by '+' or '-' then the volume is changed
relatively by that value.

=cut

sub volume {
    my ($self, $volume) = @_;

    if ($volume =~ /^(-|\+)(\d+)/ )  {
        my $current = $self->status->volume;
        $volume = $1 eq '+' ? $current + $2 : $current - $2;
    }
    $self->_send_command("setvol $volume\n");
}


=meth_mpd_output outputs

    my @outputs = $mpd->outputs( );

Return a list of C<Audio::MPD::Common::Outputs> with all outputs
available within MPD.

=cut

sub outputs {
    my ($self) = @_;

    my @lines = $self->_send_command("outputs\n");
    my (@outputs, %param);

    # parse lines in reverse order since "id" lines come first
    foreach my $line (reverse @lines) {
        my ($k,$v) = split /:\s/, $line, 2;
        $k =~ s/^output//;
        $param{$k} = $v;
        next unless $k eq 'id'; # last output param
        unshift @outputs, Audio::MPD::Common::Output->new(%param);
        %param = ();
    }

    return @outputs;
}


=meth_mpd_output output_enable

    $mpd->output_enable( $output );

Enable the specified audio output. C<$output> is the ID of the audio
output.

=cut

sub output_enable {
    my ($self, $output) = @_;
    $self->_send_command("enableoutput $output\n");
}


=meth_mpd_output output_disable

    $mpd->output_disable( $output );

Disable the specified audio output. C<$output> is the ID of the audio
output.

=cut

sub output_disable {
    my ($self, $output) = @_;
    $self->_send_command("disableoutput $output\n");
}



# -- MPD interaction: retrieving info from current state

=meth_mpd_info stats

    my $stats = $mpd->stats;

Return an L<Audio::MPD::Common::Stats> object with the current statistics
of MPD. See the associated pod for more information.

=cut

sub stats {
    my ($self) = @_;
    my %kv = $self->_cooked_command_as_kv( "stats\n" );
    return Audio::MPD::Common::Stats->new(\%kv);
}


=meth_mpd_info status

    my $status = $mpd->status;

Return an L<Audio::MPD::Common::Status> object with various information on
current MPD server settings. See the associated pod for more information.

=cut

sub status {
    my ($self) = @_;
    my %kv = $self->_cooked_command_as_kv( "status\n" );
    my $status = Audio::MPD::Common::Status->new( \%kv );
    return $status;
}


=meth_mpd_info current

    my $song = $mpd->current;

Return an L<Audio::MPD::Common::Item::Song> representing the song currently
playing.

=cut

sub current {
    my ($self) = @_;
    my ($item) = $self->_cooked_command_as_items("currentsong\n");
    return $item;
}


=meth_mpd_info song

    my $song = $mpd->song( [$song] );

Return an L<Audio::MPD::Common::Item::Song> representing the song number
C<$song>. If C<$song> is not supplied, returns the current song.

=cut

sub song {
    my ($self, $song) = @_;
    return $self->current unless defined $song;
    my ($item) = $self->_cooked_command_as_items("playlistinfo $song\n");
    return $item;
}


=meth_mpd_info songid

    my $song = $mpd->songid( [$songid] );

Return an L<Audio::MPD::Common::Item::Song> representing the song with id
C<$songid>. If C<$songid> is not supplied, returns the current song.

=cut

sub songid {
    my ($self, $songid) = @_;
    return $self->current unless defined $songid;
    my ($item) = $self->_cooked_command_as_items("playlistid $songid\n");
    return $item;
}


# -- MPD interaction: altering settings


=meth_mpd_settings repeat

    $mpd->repeat( [$repeat] );

Set the repeat mode to C<$repeat> (1 or 0). If C<$repeat> is not
specified then the repeat mode is toggled.

=cut

sub repeat {
    my ($self, $mode) = @_;

    $mode = not $self->status->repeat
        unless defined $mode; # toggle if no param
    $mode = $mode ? 1 : 0;               # force integer
    $self->_send_command("repeat $mode\n");
}


=meth_mpd_settings random

    $mpd->random( [$random] );

Set the random mode to C<$random> (1 or 0). If C<$random> is not
specified then the random mode is toggled.

=cut

sub random {
    my ($self, $mode) = @_;

    $mode = not $self->status->random
        unless defined $mode; # toggle if no param
    $mode = $mode ? 1 : 0;               # force integer
    $self->_send_command("random $mode\n");
}


=meth_mpd_settings fade

    $mpd->fade( [$seconds] );

Enable crossfading and set the duration of crossfade between songs.  If
C<$seconds> is not specified or $seconds is 0, then crossfading is
disabled.

=cut

sub fade {
    my ($self, $value) = @_;
    $value ||= 0;
    $self->_send_command("crossfade $value\n");
}


# -- MPD interaction: controlling playback

=meth_mpd_playback play

    $mpd->play( [$song] );

Begin playing playlist at song number C<$song>. If no argument supplied,
resume playing.

=cut

sub play {
    my ($self, $number) = @_;
    $number = '' unless defined $number;
    $self->_send_command("play $number\n");
}


=meth_mpd_playback playid

    $mpd->playid( [$songid] );

Begin playing playlist at song ID C<$songid>. If no argument supplied,
resume playing.

=cut

sub playid {
    my ($self, $number) = @_;
    $number ||= '';
    $self->_send_command("playid $number\n");
}


=meth_mpd_playback pause

    $mpd->pause( [$state] );

Pause playback. If C<$state> is 0 then the current track is unpaused,
if C<$state> is 1 then the current track is paused.

Note that if C<$state> is not given, pause state will be toggled.

=cut

sub pause {
    my ($self, $state) = @_;
    $state ||= ''; # default is to toggle
    $self->_send_command("pause $state\n");
}


=meth_mpd_playback stop

    $mpd->stop;

Stop playback.

=cut

sub stop {
    my ($self) = @_;
    $self->_send_command("stop\n");
}


=meth_mpd_playback next

    $mpd->next;

Play next song in playlist.

=cut

sub next {
    my ($self) = @_;
    $self->_send_command("next\n");
}


=meth_mpd_playback prev

    $mpd->prev;

Play previous song in playlist.

=cut

sub prev {
    my($self) = shift;
    $self->_send_command("previous\n");
}


=meth_mpd_playback seek

    $mpd->seek( $time, [$song]);

Seek to C<$time> seconds in song number C<$song>. If C<$song> number is
not specified then the perl module will try and seek to C<$time> in the
current song.

=cut

sub seek {
    my ($self, $time, $song) = @_;
    $time ||= 0; $time = int $time;
    $song = $self->status->song if not defined $song; # seek in current song
    $self->_send_command( "seek $song $time\n" );
}


=meth_mpd_playback seekid

    $mpd->seekid( $time, $songid );

Seek to C<$time> seconds in song ID C<$songid>. If C<$song> number is
not specified then the perl module will try and seek to C<$time> in the
current song.

=cut

sub seekid {
    my ($self, $time, $song) = @_;
    $time ||= 0; $time = int $time;
    $song = $self->status->songid if not defined $song; # seek in current song
    $self->_send_command( "seekid $song $time\n" );
}


no Moose;
__PACKAGE__->meta->make_immutable;
1;

__END__

=for Pod::Coverage
    BUILD

=head1 SYNOPSIS

    use Audio::MPD;

    my $mpd = Audio::MPD->new;
    $mpd->play;
    sleep 10;
    $mpd->next;


=head1 DESCRIPTION

L<Audio::MPD> gives a clear object-oriented interface for talking to and
controlling MPD (Music Player Daemon) servers. A connection to the MPD
server is established as soon as a new L<Audio::MPD> object is created.

Since mpd is still in 0.x versions, L<Audio::MPD> sticks to latest mpd
(0.15 as time of writing) protocol & behaviour, and does B<not> try to
maintain backward compatibility.

Note that the module will by default connect to mpd before sending any
command, and will disconnect after the command has been issued. This
scheme is far from optimal, but allows us not to care about timeout
disconnections. Because of that, the C<idle> command (new in mpd 0.14)
is B<not> (and will not) be supported in L<Audio::MPD>. This will be
implemented in L<POE::Component::Client::MPD>.

B</!\> Note that L<Audio::MPD> is using high-level, blocking sockets.
This means that if the mpd server is slow, or hangs for whatever reason,
or even crash abruptly, the program will be hung forever in this sub.
The L<POE::Component::Client::MPD> module is way safer - you're advised
to use it instead of L<Audio::MPD>. Or you can try to set C<conntype> to
C<$REUSE> (see L<Audio::MPD> constructor for more details), but you
would be then on your own to deal with disconnections.


=head2 Searching the collection

To search the collection, use the C<collection()> accessor, returning the
associated L<Audio::MPD::Collection> object. You will then be able to call:

    $mpd->collection->all_songs;

See L<Audio::MPD::Collection> documentation for more details on available
methods.


=head2 Handling the playlist

To update the playlist, use the C<playlist()> accessor, returning the
associated L<Audio::MPD::Playlist> object. You will then be able to call:

    $mpd->playlist->clear;

See L<Audio::MPD::Playlist> documentation for more details on available
methods.


=head1 SEE ALSO

You can find more information on the mpd project on its homepage at
L<http://www.musicpd.org>.wikia.com>.

Original code (2005) by Tue Abrahamsen C<< <tue.abrahamsen@gmail.com> >>,
documented in 2006 by Nicholas J. Humfrey C<< <njh@aelius.com> >>.


You can look for information on this module at:

=over 4

=item * Search CPAN

L<http://metacpan.org/release/Audio-MPD>

=item * See open / report bugs

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Audio-MPD>

=item * Mailing-list

L<http://groups.google.com/group/audio-mpd>

=item * Git repository

L<http://github.com/jquelin/audio-mpd.git>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Audio-MPD>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Audio-MPD>

=back

